c_glib/arrow-glib/compute.cpp

```diff
@@ -295,6 +328,581 @@ garrow_count_options_new(void)
+                               GError **error)
+{
+  auto arrow_array = garrow_array_get_raw(array);
+  auto memory_pool = arrow::default_memory_pool();
+  arrow::compute::FunctionContext context(memory_pool);
+  arrow::compute::Datum dictionary_encoded_datum;
+  auto status =
+    arrow::compute::DictionaryEncode(&context,
+                                     arrow::compute::Datum(arrow_array),
+                                     &dictionary_encoded_datum);
+  if (!status.ok()) {
+    std::stringstream message;
+    message << "[array][dictionary-encode] <";
+    message << arrow_array->type()->ToString();
+    message << ">";
+    garrow_error_check(error, status, message.str().c_str());
+    return NULL;
+  }
+
+  auto arrow_dictionary_encoded_array =
+    arrow::MakeArray(dictionary_encoded_datum.array());
+  auto dictionary_encoded_array =
+    garrow_array_new_raw(&arrow_dictionary_encoded_array);
+  return GARROW_DICTIONARY_ARRAY(dictionary_encoded_array);
+}
+
+/**
+ * garrow_array_count:
+ * @array: A #GArrowArray.
+ * @options: (nullable): A #GArrowCountOptions.
+ * @error: (nullable): Return location for a #GError or %NULL.
+ *
+ * Returns: The number of target values on success. If an error is occurred,
+ *   the returned value is untrustful value.
+ *
+ * Since: 0.13.0
+ */
+gint64
+garrow_array_count(GArrowArray *array,
+                   GArrowCountOptions *options,
+                   GError **error)
+{
+  auto arrow_array = garrow_array_get_raw(array);
+  auto arrow_array_raw = arrow_array.get();
+  auto memory_pool = arrow::default_memory_pool();
+  arrow::compute::FunctionContext context(memory_pool);
+  arrow::compute::Datum counted_datum;
+  arrow::Status status;
+  if (options) {
+    auto arrow_options = garrow_count_options_get_raw(options);
+    status = arrow::compute::Count(&context,
+                                   *arrow_options,
+                                   *arrow_array_raw,
+                                   &counted_datum);
+  } else {
+    arrow::compute::CountOptions arrow_options(arrow::compute::CountOptions::COUNT_ALL);
+    status = arrow::compute::Count(&context,
+                                   arrow_options,
+                                   *arrow_array_raw,
+                                   &counted_datum);
+  }
+
+  if (garrow_error_check(error, status, "[array][count]")) {
+    using ScalarType = typename arrow::TypeTraits<arrow::Int64Type>::ScalarType;
+    auto counted_scalar = std::dynamic_pointer_cast<ScalarType>(counted_datum.scalar());
+    return counted_scalar->value;
+  } else {
+    return 0;
+  }
+}
+
+/**
+ * garrow_array_count_values:
+ * @array: A #GArrowArray.
+ * @error: (nullable): Return location for a #GError or %NULL.
+ *
+ * Returns: (nullable) (transfer full):
+ *   A #GArrowStructArray of `{input type "values", int64_t "counts"}`
+ *   on success, %NULL on error.
+ *
+ * Since: 0.13.0
+ */
+GArrowStructArray *
+garrow_array_count_values(GArrowArray *array,
```

Comment:
@kou Why this function name is not `*_value_counts`?

I think `*_value_counts` is better than `*_count_values` because it can be derived trivially from the wrapped C++ function name `ValueCounts`.

Moreover, in Ruby side, `value_counts` is better for me because enumerable-statistics.gem I maintained provides `value_counts` method for Array and Enumerable.

Reply:
Because `value_counts` isn't a verb.
`compute_value_counts()` will be an alternative but it's a bit long.
PHP has `array_count_values`: http://php.net/manual/en/function.array-count-values.php

Reply:
PHP is not standard in data science field, so it cannot be a reference.
In the context of Apache Arrow, we should refer pandas.
Pandas's `value_counts` is defacto standard name in such the feature as my knowledge.
`count_values` will make programmers confusing.

Reply:
If renaming this function isn't acceptable, I'd like to propose adding `garrow_array_value_counts` as an alias.
How do you think about this?

Reply:
If someone wants pandas compatible API.
We can just provide alias as pandas compatible layer.
I think that we provide consistent API than pandas compatible API.
R uses `table` for the feature. It's different with pandas.

Reply:
I understand.

Then I have another question.  Red arrow can have pandas compatible API?

Reply:
Yes.

Reply:
OK. I'll make a pull-request.
Thank you very much!